File: ClassicalSharp/Rendering/MapRenderer.cs
// Copyright 2014-2017 ClassicalSharp | Licensed under BSD-3
using System;
using ClassicalSharp.Entities;
using ClassicalSharp.Events;
using ClassicalSharp.Map;
using ClassicalSharp.GraphicsAPI;
using ClassicalSharp.Textures;
using OpenTK;
using BlockID = System.UInt16;

namespace ClassicalSharp.Renderers {
	
	public class ChunkInfo {
		
		public ushort CentreX, CentreY, CentreZ;
		public bool Visible, Empty, PendingDelete, AllAir;
		
		public bool DrawLeft, DrawRight, DrawFront, DrawBack, DrawBottom, DrawTop;
		#if OCCLUSION
		public bool Visited = false, Occluded = false;
		public byte OcclusionFlags, OccludedFlags, DistanceFlags;
		#endif
		
		public ChunkPartInfo[] NormalParts;
		public ChunkPartInfo[] TranslucentParts;
		
		public ChunkInfo(int x, int y, int z) { Reset(x, y, z); }
		
		public void Reset(int x, int y, int z) {
			CentreX = (ushort)(x + 8);
			CentreY = (ushort)(y + 8);
			CentreZ = (ushort)(z + 8);
			
			Visible = true; Empty = false; PendingDelete = false; AllAir = false;
			DrawLeft = false; DrawRight = false; DrawFront = false;
			DrawBack = false; DrawBottom = false; DrawTop = false;
		}
	}
	
	public partial class MapRenderer : IDisposable {
		
		Game game;
		
		internal int _1DUsed = -1, chunksX, chunksY, chunksZ;
		internal int renderCount = 0;
		internal ChunkInfo[] chunks, renderChunks, unsortedChunks;
		internal bool[] usedTranslucent, usedNormal;
		internal bool[] pendingTranslucent, pendingNormal;
		internal int[] normalPartsCount, translucentPartsCount;
		internal ChunkUpdater updater;
		bool inTranslucent = false;
		
		public MapRenderer(Game game) {
			this.game = game;
			updater = new ChunkUpdater(game, this);
			SetMeshBuilder(DefaultMeshBuilder());
		}
		
		public void Dispose() { updater.Dispose(); }
		
		/// <summary> Discards any built meshes for all chunks in the map.</summary>
		public void Refresh() { updater.Refresh(); }
		
		/// <summary> Retrieves the information for the given chunk. </summary>
		public ChunkInfo GetChunk(int cx, int cy, int cz) {
			return unsortedChunks[cx + chunksX * (cy + cz * chunksY)];
		}
		
		/// <summary> Marks the given chunk as needing to be deleted. </summary>
		public void RefreshChunk(int cx, int cy, int cz) {
			if (cx < 0 || cy < 0 || cz < 0 ||
			    cx >= chunksX || cy >= chunksY || cz >= chunksZ) return;
			
			ChunkInfo info = unsortedChunks[cx + chunksX * (cy + cz * chunksY)];
			if (info.AllAir) return; // do not recreate chunks completely air
			info.Empty = false;
			info.PendingDelete = true;
		}
		
		/// <summary> Potentially generates meshes for several pending chunks. </summary>
		public void Update(double deltaTime) {
			if (chunks == null) return;
			ChunkSorter.UpdateSortOrder(game, updater);
			updater.UpdateChunks(deltaTime);
		}
		
		/// <summary> Sets the mesh builder that is used to generate meshes for chunks. </summary>
		public void SetMeshBuilder(ChunkMeshBuilder newBuilder) {
			if (updater.builder != null) updater.builder.Dispose();
			
			updater.builder = newBuilder;
			updater.builder.Init(game);
			updater.builder.OnNewMapLoaded();
		}

		/// <summary> Creates a new instance of the default mesh builder implementation. </summary>
		public ChunkMeshBuilder DefaultMeshBuilder() {
			if (game.SmoothLighting)
				return new AdvLightingMeshBuilder();
			return new NormalMeshBuilder();
		}
		
		
		/// <summary> Renders all opaque and transparent blocks. </summary>
		/// <remarks> Pixels are either treated as fully replacing existing pixel, or skipped. </remarks>
		public void RenderNormal(double deltaTime) {
			if (chunks == null) return;
			IGraphicsApi gfx = game.Graphics;
			
			int[] texIds = TerrainAtlas1D.TexIds;
			gfx.SetBatchFormat(VertexFormat.P3fT2fC4b);
			gfx.Texturing = true;
			gfx.AlphaTest = true;
			
			gfx.EnableMipmaps();
			for (int batch = 0; batch < _1DUsed; batch++) {
				if (normalPartsCount[batch] <= 0) continue;
				if (pendingNormal[batch] || usedNormal[batch]) {
					gfx.BindTexture(texIds[batch]);
					RenderNormalBatch(batch);
					pendingNormal[batch] = false;
				}
			}
			gfx.DisableMipmaps();
			
			CheckWeather(deltaTime);
			gfx.AlphaTest = false;
			gfx.Texturing = false;
			#if DEBUG_OCCLUSION
			DebugPickedPos();
			#endif
		}
		
		/// <summary> Renders all translucent (e.g. water) blocks. </summary>
		/// <remarks> Pixels drawn blend into existing geometry. </remarks>
		public void RenderTranslucent(double deltaTime) {
			if (chunks == null) return;
			IGraphicsApi gfx = game.Graphics;
			
			// First fill depth buffer
			int vertices = game.Vertices;
			gfx.SetBatchFormat(VertexFormat.P3fT2fC4b);
			gfx.Texturing = false;
			gfx.AlphaBlending = false;
			gfx.ColourWriteMask(false, false, false, false);
			for (int batch = 0; batch < _1DUsed; batch++) {
				if (translucentPartsCount[batch] <= 0) continue;
				if (pendingTranslucent[batch] || usedTranslucent[batch]) {
					RenderTranslucentBatch(batch);
					pendingTranslucent[batch] = false;
				}
			}
			game.Vertices = vertices;
			
			// Then actually draw the transluscent blocks
			gfx.AlphaBlending = true;
			gfx.Texturing = true;
			gfx.ColourWriteMask(true, true, true, true);
			gfx.DepthWrite = false; // we already calculated depth values in depth pass
			
			int[] texIds = TerrainAtlas1D.TexIds;
			gfx.EnableMipmaps();
			for (int batch = 0; batch < _1DUsed; batch++) {
				if (translucentPartsCount[batch] <= 0) continue;
				if (!usedTranslucent[batch]) continue;
				gfx.BindTexture(texIds[batch]);
				RenderTranslucentBatch(batch);
			}
			gfx.DisableMipmaps();
			
			gfx.DepthWrite = true;
			// If we weren't under water, render weather after to blend properly
			if (!inTranslucent && game.World.Env.Weather != Weather.Sunny) {
				gfx.AlphaTest = true;
				game.WeatherRenderer.Render(deltaTime);
				gfx.AlphaTest = false;
			}
			gfx.AlphaBlending = false;
			gfx.Texturing = false;
		}
		
		
		void CheckWeather(double deltaTime) {
			WorldEnv env = game.World.Env;
			Vector3 pos = game.CurrentCameraPos;
			Vector3I coords = Vector3I.Floor(pos);
			
			BlockID block = game.World.SafeGetBlock(coords);
			bool outside = coords.X < 0 || coords.Y < 0 || coords.Z < 0 || coords.X >= game.World.Width || coords.Z >= game.World.Length;
			inTranslucent = BlockInfo.Draw[block] == DrawType.Translucent
				|| (pos.Y < env.EdgeHeight && outside);

			// If we are under water, render weather before to blend properly
			if (!inTranslucent || env.Weather == Weather.Sunny) return;
			game.Graphics.AlphaBlending = true;
			game.WeatherRenderer.Render(deltaTime);
			game.Graphics.AlphaBlending = false;
		}
		
		void RenderNormalBatch(int batch) {
			IGraphicsApi gfx = game.Graphics;
			for (int i = 0; i < renderCount; i++) {
				ChunkInfo info = renderChunks[i];
				if (info.NormalParts == null) continue;

				ChunkPartInfo part = info.NormalParts[batch];
				if (part.VerticesCount == 0) continue;
				usedNormal[batch] = true;
				
				gfx.BindVb(part.VbId);
				bool drawLeft = info.DrawLeft && part.LeftCount > 0;
				bool drawRight = info.DrawRight && part.RightCount > 0;
				bool drawBottom = info.DrawBottom && part.BottomCount > 0;
				bool drawTop = info.DrawTop && part.TopCount > 0;
				bool drawFront = info.DrawFront && part.FrontCount > 0;
				bool drawBack = info.DrawBack && part.BackCount > 0;
				
				int offset = part.SpriteCount;
				if (drawLeft && drawRight) {
					gfx.FaceCulling = true;
					gfx.DrawIndexedVb_TrisT2fC4b(part.LeftCount + part.RightCount, offset);
					gfx.FaceCulling = false;
					game.Vertices += part.LeftCount + part.RightCount;
				} else if (drawLeft) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.LeftCount, offset);
					game.Vertices += part.LeftCount;
				} else if (drawRight) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.RightCount, offset + part.LeftCount);
					game.Vertices += part.RightCount;
				}
				offset += part.LeftCount + part.RightCount;
				
				if (drawFront && drawBack) {
					gfx.FaceCulling = true;
					gfx.DrawIndexedVb_TrisT2fC4b(part.FrontCount + part.BackCount, offset);
					gfx.FaceCulling = false;
					game.Vertices += part.FrontCount + part.BackCount;
				} else if (drawFront) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.FrontCount, offset);
					game.Vertices += part.FrontCount;
				} else if (drawBack) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.BackCount, offset + part.FrontCount);
					game.Vertices += part.BackCount;
				}
				offset += part.FrontCount + part.BackCount;
				
				if (drawBottom && drawTop) {
					gfx.FaceCulling = true;
					gfx.DrawIndexedVb_TrisT2fC4b(part.BottomCount + part.TopCount, offset);
					gfx.FaceCulling = false;
					game.Vertices += part.TopCount + part.BottomCount;
				} else if (drawBottom) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.BottomCount, offset);
					game.Vertices += part.BottomCount;
				} else if (drawTop) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.TopCount, offset + part.BottomCount);
					game.Vertices += part.TopCount;
				}
				
				if (part.SpriteCount == 0) continue;
				int count = part.SpriteCount / 4; // 4 per sprite
				gfx.FaceCulling = true;
				if (info.DrawRight || info.DrawFront) {
					gfx.DrawIndexedVb_TrisT2fC4b(count, 0); game.Vertices += count;
				}
				if (info.DrawLeft || info.DrawBack) {
					gfx.DrawIndexedVb_TrisT2fC4b(count, count); game.Vertices += count;
				}
				if (info.DrawLeft || info.DrawFront) {
					gfx.DrawIndexedVb_TrisT2fC4b(count, count * 2); game.Vertices += count;
				}
				if (info.DrawRight || info.DrawBack) {
					gfx.DrawIndexedVb_TrisT2fC4b(count, count * 3); game.Vertices += count;
				}
				gfx.FaceCulling = false;
			}
		}

		void RenderTranslucentBatch(int batch) {
			IGraphicsApi gfx = game.Graphics;
			for (int i = 0; i < renderCount; i++) {
				ChunkInfo info = renderChunks[i];
				if (info.TranslucentParts == null) continue;
				
				ChunkPartInfo part = info.TranslucentParts[batch];
				if (part.VerticesCount == 0) continue;
				usedTranslucent[batch] = true;
				
				gfx.BindVb(part.VbId);
				bool drawLeft = (inTranslucent || info.DrawLeft) && part.LeftCount > 0;
				bool drawRight = (inTranslucent || info.DrawRight) && part.RightCount > 0;
				bool drawBottom = (inTranslucent || info.DrawBottom) && part.BottomCount > 0;
				bool drawTop = (inTranslucent || info.DrawTop) && part.TopCount > 0;
				bool drawFront = (inTranslucent || info.DrawFront) && part.FrontCount > 0;
				bool drawBack = (inTranslucent || info.DrawBack) && part.BackCount > 0;
				
				int offset = 0;
				if (drawLeft && drawRight) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.LeftCount + part.RightCount, offset);
					game.Vertices += (part.LeftCount + part.RightCount);
				} else if (drawLeft) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.LeftCount, offset);
					game.Vertices += part.LeftCount;
				} else if (drawRight) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.RightCount, offset + part.LeftCount);
					game.Vertices += part.RightCount;
				}
				offset += part.LeftCount + part.RightCount;
				
				if (drawFront && drawBack) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.FrontCount + part.BackCount, offset);
					game.Vertices += (part.FrontCount + part.BackCount);
				} else if (drawFront) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.FrontCount, offset);
					game.Vertices += part.FrontCount;
				} else if (drawBack) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.BackCount, offset + part.FrontCount);
					game.Vertices += part.BackCount;
				}
				offset += part.FrontCount + part.BackCount;
				
				if (drawBottom && drawTop) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.BottomCount + part.TopCount, offset);
					game.Vertices += (part.BottomCount + part.TopCount);
				} else if (drawBottom) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.BottomCount, offset);
					game.Vertices += part.BottomCount;
				} else if (drawTop) {
					gfx.DrawIndexedVb_TrisT2fC4b(part.TopCount, offset + part.BottomCount);
					game.Vertices += part.TopCount;
				}
			}
		}
	}
}

File: src/Client/MapRenderer.c
#include "MapRenderer.h"
#include "Block.h"
#include "Game.h"
#include "GraphicsAPI.h"
#include "WeatherRenderer.h"
#include "World.h"
#include "Vectors.h"
#include "ChunkUpdater.h"
bool inTranslucent;

ChunkInfo* MapRenderer_GetChunk(Int32 cx, Int32 cy, Int32 cz) {
	return &MapRenderer_Chunks[MapRenderer_Pack(cx, cy, cz)];
}

void MapRenderer_RefreshChunk(Int32 cx, Int32 cy, Int32 cz) {
	if (cx < 0 || cy < 0 || cz < 0 || cx >= MapRenderer_ChunksX 
		|| cy >= MapRenderer_ChunksY || cz >= MapRenderer_ChunksZ) return;

	ChunkInfo* info = &MapRenderer_Chunks[MapRenderer_Pack(cx, cy, cz)];
	if (info->AllAir) return; /* do not recreate chunks completely air */
	info->Empty = false;
	info->PendingDelete = true;
}

void MapRenderer_Update(Real64 deltaTime) {
	if (MapRenderer_Chunks == NULL) return;
	ChunkUpdater_UpdateSortOrder();
	ChunkUpdater_UpdateChunks(deltaTime);
}

void MapRenderer_CheckWeather(Real64 deltaTime) {
	Vector3 pos = Game_CurrentCameraPos;
	Vector3I coords;
	Vector3I_Floor(&coords, &pos);

	BlockID block = World_SafeGetBlock_3I(coords);
	bool outside = coords.X < 0 || coords.Y < 0 || coords.Z < 0 || coords.X >= World_Width || coords.Z >= World_Length;
	inTranslucent = Block_Draw[block] == DRAW_TRANSLUCENT || (pos.Y < WorldEnv_EdgeHeight && outside);

	/* If we are under water, render weather before to blend properly */
	if (!inTranslucent || WorldEnv_Weather == WEATHER_SUNNY) return;
	Gfx_SetAlphaBlending(true);
	WeatherRenderer_Render(deltaTime);
	Gfx_SetAlphaBlending(false);
}

void MapRenderer_RenderNormalBatch(UInt32 batch) {
	UInt32 i;
	for (i = 0; i < MapRenderer_RenderChunksCount; i++) {
		ChunkInfo* info = MapRenderer_RenderChunks[i];
		if (info->NormalParts == NULL) continue;

		ChunkPartInfo part = info->NormalParts[batch];
		if (!part.HasVertices) continue;
		MapRenderer_HasNormalParts[batch] = true;

		Gfx_BindVb(part.VbId);
		bool drawXMin = info->DrawXMin && part.XMinCount > 0;
		bool drawXMax = info->DrawXMax && part.XMaxCount > 0;
		bool drawYMin = info->DrawYMin && part.YMinCount > 0;
		bool drawYMax = info->DrawYMax && part.YMaxCount > 0;
		bool drawZMin = info->DrawZMin && part.ZMinCount > 0;
		bool drawZMax = info->DrawZMax && part.ZMaxCount > 0;

		UInt32 offset = part.SpriteCountDiv4 << 2;
		if (drawXMin && drawXMax) {
			Gfx_SetFaceCulling(true);
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMinCount + part.XMaxCount, offset);
			Gfx_SetFaceCulling(false);
			Game_Vertices += part.XMinCount + part.XMaxCount;
		} else if (drawXMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMinCount, offset);
			Game_Vertices += part.XMinCount;
		} else if (drawXMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMaxCount, offset + part.XMinCount);
			Game_Vertices += part.XMaxCount;
		}
		offset += part.XMinCount + part.XMaxCount;

		if (drawZMin && drawZMax) {
			Gfx_SetFaceCulling(true);
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMinCount + part.ZMaxCount, offset);
			Gfx_SetFaceCulling(false);
			Game_Vertices += part.ZMinCount + part.ZMaxCount;
		} else if (drawZMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMinCount, offset);
			Game_Vertices += part.ZMinCount;
		} else if (drawZMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMaxCount, offset + part.ZMinCount);
			Game_Vertices += part.ZMaxCount;
		}
		offset += part.ZMinCount + part.ZMaxCount;

		if (drawYMin && drawYMax) {
			Gfx_SetFaceCulling(true);
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMinCount + part.YMaxCount, offset);
			Gfx_SetFaceCulling(false);
			Game_Vertices += part.YMaxCount + part.YMinCount;
		} else if (drawYMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMinCount, offset);
			Game_Vertices += part.YMinCount;
		} else if (drawYMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMaxCount, offset + part.YMinCount);
			Game_Vertices += part.YMaxCount;
		}

		if (part.SpriteCountDiv4 == 0) continue;
		UInt32 count = part.SpriteCountDiv4; /* 4 per sprite */
		Gfx_SetFaceCulling(true);
		if (info->DrawXMax || info->DrawZMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(count, 0); Game_Vertices += count;
		}
		if (info->DrawXMin || info->DrawZMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(count, count); Game_Vertices += count;
		}
		if (info->DrawXMin || info->DrawZMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(count, count * 2); Game_Vertices += count;
		}
		if (info->DrawXMax || info->DrawZMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(count, count * 3); Game_Vertices += count;
		}
		Gfx_SetFaceCulling(false);
	}
}

void MapRenderer_RenderNormal(Real64 deltaTime) {
	if (MapRenderer_Chunks == NULL) return;
	Gfx_SetBatchFormat(VERTEX_FORMAT_P3FT2FC4B);
	Gfx_SetTexturing(true);
	Gfx_SetAlphaTest(true);

	UInt32 batch;
	Gfx_EnableMipmaps();
	for (batch = 0; batch < MapRenderer_1DUsedCount; batch++) {
		if (MapRenderer_NormalPartsCount[batch] <= 0) continue;
		if (MapRenderer_HasNormalParts[batch] || MapRenderer_CheckingNormalParts[batch]) {
			Gfx_BindTexture(Atlas1D_TexIds[batch]);
			MapRenderer_RenderNormalBatch(batch);
			MapRenderer_CheckingNormalParts[batch] = false;
		}
	}
	Gfx_DisableMipmaps();

	MapRenderer_CheckWeather(deltaTime);
	Gfx_SetAlphaTest(false);
	Gfx_SetTexturing(false);
#if DEBUG_OCCLUSION
	DebugPickedPos();
#endif
}

void MapRenderer_RenderTranslucentBatch(UInt32 batch) {
	UInt32 i;
	for (i = 0; i < MapRenderer_RenderChunksCount; i++) {
		ChunkInfo* info = MapRenderer_RenderChunks[i];
		if (info->TranslucentParts == NULL) continue;

		ChunkPartInfo part = info->TranslucentParts[batch];
		if (!part.HasVertices) continue;
		MapRenderer_HasTranslucentParts[batch] = true;

		Gfx_BindVb(part.VbId);
		bool drawXMin = (inTranslucent || info->DrawXMin) && part.XMinCount > 0;
		bool drawXMax = (inTranslucent || info->DrawXMax) && part.XMaxCount > 0;
		bool drawYMin = (inTranslucent || info->DrawYMin) && part.YMinCount > 0;
		bool drawYMax = (inTranslucent || info->DrawYMax) && part.YMaxCount > 0;
		bool drawZMin = (inTranslucent || info->DrawZMin) && part.ZMinCount > 0;
		bool drawZMax = (inTranslucent || info->DrawZMax) && part.ZMaxCount > 0;

		UInt32 offset = 0;
		if (drawXMin && drawXMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMinCount + part.XMaxCount, offset);
			Game_Vertices += (part.XMinCount + part.XMaxCount);
		} else if (drawXMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMinCount, offset);
			Game_Vertices += part.XMinCount;
		} else if (drawXMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.XMaxCount, offset);
			Game_Vertices += part.XMaxCount;
		}
		offset += part.XMinCount + part.XMaxCount;

		if (drawZMin && drawZMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMinCount + part.ZMaxCount, offset);
			Game_Vertices += (part.ZMinCount + part.ZMaxCount);
		} else if (drawZMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMinCount, offset);
			Game_Vertices += part.ZMinCount;
		} else if (drawZMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.ZMaxCount, offset + part.ZMinCount);
			Game_Vertices += part.ZMaxCount;
		}
		offset += part.ZMinCount + part.ZMaxCount;

		if (drawYMin && drawYMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMinCount + part.YMaxCount, offset);
			Game_Vertices += (part.YMinCount + part.YMaxCount);
		} else if (drawYMin) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMinCount, offset);
			Game_Vertices += part.YMinCount;
		} else if (drawYMax) {
			Gfx_DrawIndexedVb_TrisT2fC4b(part.YMaxCount, offset + part.YMinCount);
			Game_Vertices += part.YMaxCount;
		}
	}
}

void MapRenderer_RenderTranslucent(Real64 deltaTime) {
	if (MapRenderer_Chunks == NULL) return;

	/* First fill depth buffer */
	UInt32 vertices = Game_Vertices;
	Gfx_SetBatchFormat(VERTEX_FORMAT_P3FT2FC4B);
	Gfx_SetTexturing(false);
	Gfx_SetAlphaBlending(false);
	Gfx_SetColourWriteMask(false, false, false, false);

	UInt32 batch;
	for (batch = 0; batch < MapRenderer_1DUsedCount; batch++) {
		if (MapRenderer_TranslucentPartsCount[batch] <= 0) continue;
		if (MapRenderer_HasTranslucentParts[batch] || MapRenderer_CheckingTranslucentParts[batch]) {
			MapRenderer_RenderTranslucentBatch(batch);
			MapRenderer_CheckingTranslucentParts[batch] = false;
		}
	}
	Game_Vertices = vertices;

	/* Then actually draw the transluscent blocks */
	Gfx_SetAlphaBlending(true);
	Gfx_SetTexturing(true);
	Gfx_SetColourWriteMask(true, true, true, true);
	Gfx_SetDepthWrite(false); /* we already calculated depth values in depth pass */

	Gfx_EnableMipmaps();
	for (batch = 0; batch < MapRenderer_1DUsedCount; batch++) {
		if (MapRenderer_TranslucentPartsCount[batch] <= 0) continue;
		if (!MapRenderer_HasTranslucentParts[batch]) continue;
		Gfx_BindTexture(Atlas1D_TexIds[batch]);
		MapRenderer_RenderTranslucentBatch(batch);
	}
	Gfx_DisableMipmaps();

	Gfx_SetDepthWrite(true);
	/* If we weren't under water, render weather after to blend properly */
	if (!inTranslucent && WorldEnv_Weather != WEATHER_SUNNY) {
		Gfx_SetAlphaTest(true);
		WeatherRenderer_Render(deltaTime);
		Gfx_SetAlphaTest(false);
	}
	Gfx_SetAlphaBlending(false);
	Gfx_SetTexturing(false);
}